In [25]:
# Import libraries
# import pandas as pd # <-- Not needed
import sklearn
from sklearn import datasets
from sklearn.model_selection import train_test_split

In [33]:
# Loading the dataset
# For summary purpose the as_frame argument is going to be set to true 
# to convert it to a pandas dataframe-like without having to load the pandas library

iris = sklearn.datasets.load_iris(return_X_y=False, as_frame=True)
df = iris.frame
df.head(10) # <-- view a summary of the loaded data
# print("Load 10 rows summary", iris.data[:10]) # <-- This could also be achived this way

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm),target
0,5.1,3.5,1.4,0.2,0
1,4.9,3.0,1.4,0.2,0
2,4.7,3.2,1.3,0.2,0
3,4.6,3.1,1.5,0.2,0
4,5.0,3.6,1.4,0.2,0
5,5.4,3.9,1.7,0.4,0
6,4.6,3.4,1.4,0.3,0
7,5.0,3.4,1.5,0.2,0
8,4.4,2.9,1.4,0.2,0
9,4.9,3.1,1.5,0.1,0


## Understanding the Dataset

In [36]:
# View a statistical summary of the dataset
df.describe()

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm),target
count,150.000000,150.000000,150.000000,150.000000,150.000000
mean,5.843333,3.057333,3.758000,1.199333,1.000000
std,0.828066,0.435866,1.765298,0.762238,0.819232
min,4.300000,2.000000,1.000000,0.100000,0.000000
25%,5.100000,2.800000,1.600000,0.300000,0.000000
50%,5.800000,3.000000,4.350000,1.300000,1.000000
75%,6.400000,3.300000,5.100000,1.800000,2.000000
max,7.900000,4.400000,6.900000,2.500000,2.000000


In [42]:
df["target"].value_counts()

target
0    50
1    50
2    50
Name: count, dtype: int64

In [49]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
 4   target             150 non-null    int32  
dtypes: float64(4), int32(1)
memory usage: 5.4 KB


In [53]:
print("Features names: ", iris.feature_names)
print("Name of real species: ", iris.target_names)

Features names:  ['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']
Name of real species:  ['setosa' 'versicolor' 'virginica']


1. This dataset has three _classes_ for each iris species. 
2. Each class has 50 _samples_ per class.
3. The total sample is 150.
4. Most features are real, postive (with datatypes: float and int)
5. Features names:  ['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']
6. The datasize in memory is 5.4 KB.


From the overall summary we can see that target contains real species data which are strongly affected by
the features measures. 
In this case, target is a dependent variable and features are independet variables.

### --> Proceed with the dataset split into test and training
1. X_features (_independet variables_) contains all features except, target
2. y_target (_dependet variable_), target only

In [66]:
# Use the same split for fair comparison between models.
# Note that the choosen test-size is 30% which will potentially provide more accuracy

# define X_, and
X_features = iris.frame.drop(columns=['target'])
y_target = iris.frame['target']

X_train, X_test, y_train, y_test = train_test_split(
    X_features, 
    y_target, 
    test_size=0.30, 
    random_state=42
)

## Build a Tecision Tree

In [70]:
# Import required libraries

from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

# Initialize the decision classifier
tree_classifier = DecisionTreeClassifier(random_state=42)

# Training the model with the 70% of data provided
tree_classifier.fit(X_train, y_train)

# Make prediction of the 30% of test data
y_prediction_tree = tree_classifier.predict(X_test)

# Evaluate the model accuracy
accuracy_tree = accuracy_score(y_test, y_prediction_tree)
print(f"The Precision of the Decision Tree is: {accuracy_tree:.2%}")

The Precision of the Decision Tree is: 100.00%


# Build an Rule-Based Classifier

In [72]:
# Import required libraries
import pandas as pd

def predict_by_rules(features_dataframe: pd.DataFrame) -> list:
    """Manual classiffier based on predefined business rules"""
    rule_predictions = []
    
    for _, row in features_dataframe.iterrows():
        petal_length = row['petal length (cm)']
        petal_width = row['petal width (cm)']
        
        # Rule 1: If petals is very short, its cassified as Setosa (0)
        if petal_length < 2.5:
            rule_predictions.append(0)
        # Rule 2: If petals has avg demensions, is Versicolor (1)
        elif 2.5 <= petal_length < 4.8 and petal_width < 1.7:
            rule_predictions.append(1)
        # Regla 3: If the dimensions are greater, is Virginica (2)
        else:
            rule_predictions.append(2)
            
    return rule_predictions

# Predition generation utilizing ruled based logic
y_pred_rules = predict_by_rules(X_test)

# Model evaluation
accuracy_rules = accuracy_score(y_test, y_pred_rules)
print(f"The Ruled-Based Classiffier is: {accuracy_rules:.2%}")


The Ruled-Based Classiffier is: 97.78%


## Build a Naive Bayes Classifier

In [74]:
# import required libraries
from sklearn.naive_bayes import GaussianNB

# Initialized probabilistic Naive Bayes model
naive_bayes_classifier = GaussianNB()

# Train the model
naive_bayes_classifier.fit(X_train, y_train)

# Predict over testing data
y_pred_naive_bayes = naive_bayes_classifier.predict(X_test)

# Evaluate model
accuracy_naive_bayes = accuracy_score(y_test, y_pred_naive_bayes)
print(f"Naive Bayes Precision: {accuracy_naive_bayes:.2%}")

Naive Bayes Precision: 97.78%


### Test the model

In [100]:
# To test the model predicted result, we create a new registry in iris
# If the new flower has the followign: sepals (4.9x3.2 cm) and petals (1.8x0.6 cm)..
new_data = pd.DataFrame([
    {'sepal length (cm)': 3.9,'sepal width (cm)': 4.2,'petal length (cm)': 1.8,'petal width (cm)': 1.9},
    {'sepal length (cm)': 4.9,'sepal width (cm)': 3.2,'petal length (cm)': 2.8,'petal width (cm)': 0.6},
])

# Translating the species name to targeted numbers
species_names = iris.target_names

print("--- INDIVIDUAL PREDICTION RESULTS ---")

# Test 1: Decision Tree
numeric_tree_prediction = tree_classifier.predict(new_data)
tree_species = species_names[numeric_tree_prediction[0]]
print(f"Decision Tree outputs: {tree_species}")

# Test 2: Ruled-Based Classifier
numeric_rules_prediction = predict_by_rules(new_data)
rules_species = species_names[numeric_rules_prediction[0]]
print(f"Rule-Based outputs:  {rules_species}")

# Test 3: Naive Bayes
numeric_pridiction_nb = naive_bayes_classifier.predict(new_data)
nb_species = species_names[numeric_pridiction_nb[0]]
print(f"Naive Bayes outputs: {nb_species}")

print("\n--- MULTIPLE PREDICTION RESULTS ---")

multiple_pridictions = tree_classifier.predict(new_data)
print("Mass Predictions (IDs):", multiple_pridictions)
print("Pridiction by Species:", [species_names[i] for i in multiple_pridictions])

--- INDIVIDUAL PREDICTION RESULTS ---
Decision Tree outputs: setosa
Rule-Based outputs:  setosa
Naive Bayes outputs: versicolor

--- MUTIPLE PREDICTION RESULTS ---
Mass Predictions (IDs): [0 1]
Pridiction by Species: ['setosa', 'versicolor']


### Build and Test the following three Distance and Optimization-Based Models: 
1. Logistic Regression
2. K-Nearest Neighbors (KNN)
3. Support Vector Machine (SVM)

In [119]:
# import only required libraries
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score
import pandas as pd

# Reload the dataset to prevent inaccurate predition from the above already tested data
iris = sklearn.datasets.load_iris(as_frame=True)
X_features = iris.frame.drop(columns=['target'])
y_target = iris.frame['target']

# Implement a scaler with Train data and applied to Test and Train
# This will transform the data with a median of 0 and variance of 1
# necessary for standardScaler

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

### Build a Logistic Regression

In [121]:
# import required library
from sklearn.linear_model import LogisticRegression

# Initialized and train the classifier
logistic_regression_classifier = LogisticRegression(random_state=42)
logistic_regression_classifier.fit(X_train_scaled, y_train)

# Predict on the testing data
y_pred_logistic = logistic_regression_classifier.predict(X_test_scaled)

# Evaluate the predition
accuracy_logistic = accuracy_score(y_test, y_pred_logistic)
print(f"Precisión de Regresión Logística: {accuracy_logistic:.2%}")

Precisión de Regresión Logística: 100.00%


### Build a K-Nearest Neighbors (KNN)

In [123]:
# Import required libraries
from sklearn.neighbors import KNeighborsClassifier

# Initialize the data using 5 neighbors by default and train 
knn_classifier = KNeighborsClassifier(n_neighbors=5)
knn_classifier.fit(X_train_scaled, y_train)

# Predict on the testing data
y_pred_knn = knn_classifier.predict(X_test_scaled)

# Evaluate the prediction
accuracy_knn = accuracy_score(y_test, y_pred_knn)
print(f"Precisión de K-Nearest Neighbors:  {accuracy_knn:.2%}")

Precisión de K-Nearest Neighbors:  100.00%


### Build a Support Vector Machine

In [125]:
# import required libraries
from sklearn.svm import SVC

# Initialize the data using the RBF kernel by default and train
svm_classifier = SVC(kernel='rbf', random_state=42)
svm_classifier.fit(X_train_scaled, y_train)

# Predict on the testing data
y_pred_svm = svm_classifier.predict(X_test_scaled)

# Evaluate the predictions
accuracy_svm = accuracy_score(y_test, y_pred_svm)
print(f"Precisión de Support Vector Machine: {accuracy_svm:.2%}")

Precisión de Support Vector Machine: 100.00%


### Reuse the testing data from above

In [142]:
# To test the model predicted result, we create a new registry in iris
# If the new flower has the followign: sepals (4.9x3.2 cm) and petals (1.8x0.6 cm)..
new_data = pd.DataFrame([
    {'sepal length (cm)': 3.9, 'sepal width (cm)': 4.2, 'petal length (cm)': 1.8, 'petal width (cm)': 1.9},
    {'sepal length (cm)': 4.9, 'sepal width (cm)': 3.2, 'petal length (cm)': 5.1, 'petal width (cm)': 0.6},
])

# Scale to the new data
new_scaled_data = scaler.transform(new_data)

# Translating the species name to targeted numbers
species_names = iris.target_names

print("--- TESTING DISTANCE AND OPTIMATION ALGORITHMS ---")

print(f" Logistic Regression Outputs: {species_names[logistic_regression_classifier.predict(new_scaled_data)[0]]}")
print(f" K-Nearest Neighbors (KNN) Outputs: {species_names[knn_classifier.predict(new_scaled_data)[0]]}")
print(f"Support Vector Machine (SVM) Outputs: {species_names[svm_classifier.predict(new_scaled_data)[0]]}")

print("\n--- MULTIPLE PREDICTION RESULTS ---")

multiple_pridictions = tree_classifier.predict(new_data)
print("Mass Predictions (IDs):", multiple_pridictions)
print("Pridiction by Species:", [species_names[i] for i in multiple_pridictions])

--- TESTING DISTANCE AND OPTIMATION ALGORITHMS ---
 Logistic Regression Outputs: setosa
 K-Nearest Neighbors (KNN) Outputs: setosa
Support Vector Machine (SVM) Outputs: setosa

--- MULTIPLE PREDICTION RESULTS ---
Mass Predictions (IDs): [0 2]
Pridiction by Species: ['setosa', 'virginica']


### Evaluate each model using the following three metrics: 
1. Confusion Matrix
2. Precision and Recall
3. F1-score  

In [147]:
# import required libraries
from sklearn.metrics import confusion_matrix, precision_score, recall_score, f1_score

# Create a dictionary to store the species names
target_names = iris.target_names

# Model information list with respective predictions
models_evaluation = [
    {"name": "Regresión Logística", "predictions": y_pred_logistic},
    {"name": "K-Nearest Neighbors (KNN)", "predictions": y_pred_knn},
    {"name": "Support Vector Machine (SVM)","predictions": y_pred_svm}
]

print("=== Model Evaluation Report ===\n")

for model in models_evaluation:
    name = model["name"]
    y_pred = model["predictions"]
    
    # 1. Confusion Matrix Calculation
    cm = confusion_matrix(y_test, y_pred)
    
    # 2. individual metrics calcualtions
    precision = precision_score(y_test, y_pred, average='weighted')
    recall = recall_score(y_test, y_pred, average='weighted')
    f1 = f1_score(y_test, y_pred, average='weighted')
    
    # print scalable results
    print(f"MODELO: {name}")
    print("-" * 40)
    
    print("CONFUSION MATRIX:")
    # Matrx formating
    cm_df = pd.DataFrame(cm, index=[f"Real {n}" for n in target_names], columns=[f"Pred {n}" for n in target_names])
    print(cm_df)
    print("")
    
    print(f" Precision (Precision):  {precision:.2%}")
    print(f" Recall (Recall):     {recall:.2%}")
    print(f" F1-Score (F1-score): {f1:.2%}")
    print("\n" + "="*50 + "\n")


=== Model Evaluation Report ===

MODELO: Regresión Logística
----------------------------------------
CONFUSION MATRIX:
                 Pred setosa  Pred versicolor  Pred virginica
Real setosa               19                0               0
Real versicolor            0               13               0
Real virginica             0                0              13

 Precision (Precision):  100.00%
 Recall (Recall):     100.00%
 F1-Score (F1-score): 100.00%


MODELO: K-Nearest Neighbors (KNN)
----------------------------------------
CONFUSION MATRIX:
                 Pred setosa  Pred versicolor  Pred virginica
Real setosa               19                0               0
Real versicolor            0               13               0
Real virginica             0                0              13

 Precision (Precision):  100.00%
 Recall (Recall):     100.00%
 F1-Score (F1-score): 100.00%


MODELO: Support Vector Machine (SVM)
----------------------------------------
CONFUSION MATRIX:
   